src/reducers/cacheReducer.js

```diff
@@ -232,6 +241,73 @@ const filterTransducers = (where) => {
     );
   });
 };
+
+/**
+ * @name paginateTransducers
+ * @param {RRFQuery} query - Firestore wquery
+ * @typedef {Function} xFormFilter - run the same where cause sent to
+ * Only apply pagination if we get the full list of docs, pre-sorted
+ * @returns {xFormFilter} - transducer
+ */
+const paginateTransducers = (query) => {
+  const { orderBy: order, startAt, startAfter, endAt, endBefore, via } = query;
+  const isOptimisticRead = via === undefined;
+  if (!isOptimisticRead) return null;
+
+  const start = startAt || startAfter;
+  const end = endAt || endBefore;
+  const isAfter = startAfter !== undefined;
+  const isBefore = endBefore !== undefined;
+  if (start === undefined && end === undefined) return null;
+
+  const isFlat = typeof order[0] === 'string';
+  const orders = isFlat ? [order] : order;
+  const isPaginateMatched = (doc, at, before = false, after = false) => {
+    let matched = null;
+    orders.forEach(([field, sort = 'asc'], idx) => {
+      if (matched !== null) return;
+      const value = Array.isArray(at) ? at[idx] : at;
+      if (value === undefined) return;
+
+      // TODO: add support for document refs
+      const proc = isTimestamp(doc[field]) ? PROCESSES_TIMESTAMP : PROCESSES;
+      let compare = process['=='];
+      if (startAt || endAt) compare = proc[sort === 'desc' ? '<=' : '>='];
+      if (startAfter || endBefore) compare = proc[sort === 'desc' ? '<' : '>'];
+
+      const isMatched = compare(doc[field], value);
+      if (isMatched) {
+        matched = true;
+      }
+    });
+
+    return matched;
+  };
+
+  return partialRight(map, (docs) => {
+    const results = [];
+    let started = start === undefined;
+
+    docs.forEach((doc, idx) => {
+      if (!started && start) {
+        const matched = isPaginateMatched(doc, start, undefined, isAfter);
+        if (matched !== null) started = true;
```

Comment:
I would not declare twice the same `const`.
```suggestion
        started = (isPaginateMatched(doc, start, undefined, isAfter) !== null);
```

Reply:
switched to boolean so it's easier.